### PR TITLE
Enhance directory traversal handling/resolution.

### PR DESCRIPTION
**Pull request by barryhughes**
### Changes proposed in this Pull Request:

This set of changes is a necessary precursor to some other changes that are required to support shortcodes as a means of supplying a downloadable file URL (as is the case with the S3 downloads add-on).

Specifically, it guards against tricky issues that might occur if a plugin like S3 is disabled and/or other scenarios where a disused shortcode is supplied to represent a downloadable filepath or URL (because in those cases, the shortcode is  unparseable and essentially represents a relative filepath.

Additionally, it improves logic relating to handling of filepaths and URLs that contain directory traversals from within our internally-facing URL utility class.

### How to test the changes in this Pull Request:

There is no UI-driven way to assess this set of changes. Instead, please review the newly added tests, and revisions to the existing tests. Those are probably the best indication of the scenarios I wanted to touch on.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

Not needed.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Utilities/URL.php [hunk 1]
```diff
@@ -124,8 +131,11 @@ private function preprocess() {
 	 * without touching the filesystem.
 	 */
 	private function process_path() {
-		$segments          = explode( '/', $this->components['path'] );
-		$this->is_absolute = substr( $this->components['path'], 0, 1 ) === '/';
+		$segments           = explode( '/', $this->components['path'] );
+		$this->is_absolute  = substr( $this->components['path'], 0, 1 ) === '/' || ! empty( $this->components['host'] );
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Utilities/URL.php]: does `absolute` here mean pointing to a specific file (as opposed to a relative URL or file)? if so, then what would be the case where `! empty( $this->components['host'] )` would also be needed (just curious)?

File: plugins/woocommerce/src/Internal/Utilities/URL.php [hunk 2]
```diff
@@ -124,8 +131,11 @@ private function preprocess() {
...
+		$this->is_directory = substr( $this->components['path'], -1, 1 ) === '/' && strlen( $this->components['path'] ) > 1;
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Utilities/URL.php]: why `strlen( $this->components['path'] ) > 1;`? just `/` is also a directory right?

File: plugins/woocommerce/src/Internal/Utilities/URL.php [hunk 3]
```diff
@@ -137,19 +147,43 @@ private function process_path() {
 			// Directory traversals created with percent-encoding syntax should also be detected.
 			$is_traversal = str_ireplace( '%2e', '.', $part ) === '..';
 
-			// Unwind directory traversals.
-			if ( $is_traversal && count( $this->path_parts ) > 0 ) {
-				$this->path_parts = array_slice( $this->path_parts, 0, count( $this->path_parts ) - 1 );
-				continue;
+			// Resolve directory traversals (if allowed: see further comment relating to this).
+			if ( $resolve_traversals && $is_traversal ) {
+				if ( count( $this->path_parts ) > 0 && ! $retain_traversals ) {
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Utilities/URL.php:156]: there might be an edge case when the path starts from `./`, for example, `./.././../foo.txt`

Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/Utilities/URL.php:156]: what do we do when the path starts from other special chars, such as `~`, `-` etc?

File: plugins/woocommerce/tests/php/src/Internal/Utilities/URLTest.php [hunk 4]
```diff
@@ -42,7 +42,19 @@ public function test_directory_traversal_resolution() {
...
+			( new URL( 'relative/../../../should/remain/relative' ) )->get_path(),
+			'Simplifies a relative path containing directory traversals to the extent possible (without inspecting the filesystem - scenario #2).'
+		);
+
+		$this->assertEquals(
+			'file:///foo/bar/baz',
+			( new URL( '/../foo/bar/baz/bazooka/../../baz' ) )->get_url(),
+			'Directory traversals are appropriately resolved even in complex cases with multiple separate traversals. When the original path is absolute, the output will be absolute.'
 		);
 	}
```
Comment (vedanshujain) [inline plugins/woocommerce/tests/php/src/Internal/Utilities/URLTest.php:38]: Let's add more tests:
`./.././.././../foo.txt`
`~/foo.txt`
`///foo`
`bazz//foo`
`foo://bar`
`http://localhost?../../bar`

**Review by vedanshujain — COMMENTED** [review]
Added some preliminary comments, but still testing in 32362

File: plugins/woocommerce/src/Internal/Utilities/URL.php (hunk 1 above)
Comment (barryhughes) [reply plugins/woocommerce/src/Internal/Utilities/URL.php]: Yes, in the case of file URLs we are checking if it represents an absolute filepath, whether that is a directory or a file. 

For URLs with host components (generally, anything that is not a file URL), we also want to treat them as being absolute specifically in relation to how any leading directory traversals are handled.

The inputs and outputs we want (simplified, but covers this aspect):

| Type | Unprocessed (Host+)Path | Processed Path |
| --- | --- | --- |
| **Absolute** | `/../../remainder` | `/remainder` |
| **Absolute** | `example.com/../../remainder` | `example.com/remainder` |
| **Relative** | `../../remainder` | `../../remainder` |

File: plugins/woocommerce/src/Internal/Utilities/URL.php (hunk 2 above)
Comment (barryhughes) [reply plugins/woocommerce/src/Internal/Utilities/URL.php]: Yes, it is also a directory; I'll think about a different name as `$is_directory` is perhaps a little misleading (as to why we do this, one of the reasons is that when we reconstitute the path at the end of processing, we don't want to apply an extra trailing slash to root dir `/` (ie, we don't want `//`).

File: plugins/woocommerce/src/Internal/Utilities/URL.php (hunk 3 above)
Comment (barryhughes) [reply plugins/woocommerce/src/Internal/Utilities/URL.php:156]: Both are treated as any other character (and, though `~` has special meaning in filepaths, I don't believe `-` does?).

I'll look at documenting what we expect for those cases (and the possible edge case) via further unit tests, though, along with the others you suggest further down :+1:

**Comment by barryhughes** [comment]
Updated (and rebased following PHP 7.2-related changes that were merged to trunk).

@vedanshujain can you take another look? I also switched to using data providers in the tests, I felt this was more readable and it makes it fast and easy to drop-in additional scenarios if needed.

**Review by vedanshujain — APPROVED** [review]
Looks pretty neat!

File: plugins/woocommerce/src/Internal/Utilities/URL.php (hunk 3 above)
Comment (barryhughes) [reply plugins/woocommerce/src/Internal/Utilities/URL.php:156]: This note doesn't necessarily require a follow-up, I just thought there was a chance it may be handy to record a note someplace: though of course `~` at the start of a path is expanded by most shells to the current user home dir, it is not treated that way when working with `fopen()` and related functions: in those cases, it is just another symbol (for instance, if the current working dir is `/tmp`, then `~/file.txt` would mean `/tmp/~/file.txt` ... an unusual but valid path).

**Comment by barryhughes** [comment]
(Changelog/further testing instructions not required: this is an amendment to an as-yet-unreleased change.)
